08 - Fun with HTML5 Canvas/index-CLASS.html

```diff
@@ -0,0 +1,128 @@
+                this.context = this.canvas.getContext(contextID)
+                this.X = 0;
+                this.Y = 0;
+                this.hue = 0;
+                this.isDrawing = false;
+                this.isStrokeGrowing = true;
+
+                //// binding this to events
+                this.startDrawing = this.startDrawing.bind(this);
+                this.stopDrawing = this.stopDrawing.bind(this);
+                this.clearCanvas = this.clearCanvas.bind(this);
+                this.draw = this.draw.bind(this);
+            }
+
+            setCanvas() {
+                this.canvas.width = window.innerWidth;
+                this.canvas.height = window.innerHeight;
+                this.context.strokeStyle = '#000';
+                this.context.lineJoin = 'miter';
+                this.context.lineCap = 'round';
+                this.context.lineWidth = 1;
+            }
+
+            updateCoordinates() {
+                this.X = this.offsetX;
+                this.Y = this.offsetY;
+            };
+
+            startDrawing() {
+                this.isDrawing = true;
+            };
+
+            stopDrawing() {
+                this.isDrawing = false;
+            };
+
+            clearCanvas() {
+                // Note: make sure to use beginPath() when starting the new line in draw() or clearRect() could give side effects
+                this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
+                this.context.lineWidth = 1;
+                this.hue = 0;
+            };
+
+            draw(evt) {
+                if (!this.isDrawing) return;
+                
+                // Note: do not remove beginPath() due to side effects of clearRect() in clearCanvas()
+                this.context.strokeStyle = `hsl(${this.hue}, 100%, 50%)`;
+                this.context.beginPath();
+                this.context.moveTo(this.X, this.Y);
+                this.context.lineTo(evt.offsetX, evt.offsetY);
+                this.context.stroke();
+                
+                this.X = evt.offsetX;
+                this.Y = evt.offsetY;
+                
+                this.hue++;
+                if (this.hue >= 360) this.hue = 0;
+                if (this.context.lineWidth >= 100 || this.context.lineWidth <= 1) this.isStrokeGrowing = !this.isStrokeGrowing;
+                if (this.isStrokeGrowing) this.context.lineWidth++;
+                else this.context.lineWidth--;
+            };
+        };
+
+        //// selectors
+        const button = document.querySelector('.clearCanvas');
+        
+        //// init
+        const myCanvas = new DrawingCanvas('#draw');
+        myCanvas.setCanvas();
+
+        //// listeners
+        myCanvas.canvas.addEventListener('mousedown', () => {
+            myCanvas.startDrawing();
+            myCanvas.updateCoordinates();
+        });
+        myCanvas.canvas.addEventListener('mousemove', myCanvas.draw);
```

Comment:
who's responsibility is it to setup the event listeners? the class's or the developers? can you think of ways why it may be useful to put this burden on the developer?

Reply:
I think that the class should provide methods, which are flexible enough to be called by the developer when needed. For example the developer could decide to code the same canvas features but using arrow keys up/down etc... or on different devices which do not support some events.